wandb/sdk/launch/registry/google_artifact_registry.py

```diff
@@ -0,0 +1,118 @@
+    image_name: str
+    environment: GcpEnvironment
+
+    def __init__(
+        self,
+        repository: str,
+        image_name: str,
+        environment: GcpEnvironment,
+        verify: bool = True,
+    ) -> None:
+        """Initialize the Google Artifact Registry.
+
+        Args:
+            repository: The repository name.
+            image_name: The image name.
+            environment: A GcpEnvironment configured for access to this registry.
+            verify: Whether to verify the credentials, region, and project.
+
+        Raises:
+            LaunchError: If verify is True and the container registry or its
+                environment have not been properly configured.
+        """
+        _logger.info(
+            f"Initializing Google Artifact Registry with repository {repository} "
+            f"and image name {image_name}"
+        )
+        self.repository = repository
+        self.image_name = image_name
+        if not re.match(r"^\w[\w.-]+$", image_name):
+            raise LaunchError(
+                f"The image name {image_name} is invalid. The image name must "
+                "consist of alphanumeric characters and underscores."
+            )
+        self.environment = environment
+        if verify:
+            self.verify()
+
+    def verify(self) -> None:
+        """Verify the registry is properly configured.
+
+        Raises:
+            LaunchError: If the registry is not properly configured.
+        """
+        credentials = self.environment.get_credentials()
+        parent = (
+            f"projects/{self.environment.project}/locations/{self.environment.region}"
+        )
+        # We need to list the repositories to verify that the repository exists.
+        request = google.cloud.artifactregistry.ListRepositoriesRequest(parent=parent)
+        client = google.cloud.artifactregistry.ArtifactRegistryClient(
+            credentials=credentials
+        )
+        try:
+            response = client.list_repositories(request=request)
+        except google.api_core.exceptions.PermissionDenied:
+            raise LaunchError(
+                "The provided credentials do not have permission to access the "
+                f"Google Artifact Registry repository {self.repository}."
+            )
+        # Look for self.repository in the list of responses.
+        for repo in response:
+            if repo.name.endswith(self.repository):
+                break
+        # If we didn't find the repository, raise an error.
+        else:
+            raise LaunchError(
+                f"The Google Artifact Registry repository {self.repository} does not exist."
+            )
+
+    def get_username_password(self) -> Tuple[str, str]:
+        """Get the username and password for the registry.
+
+        Returns:
+            A tuple of the username and password.
+        """
+        credentials = self.environment.get_credentials()
+        return "_token", credentials.token
```

Comment:
How do we check whether an image exists in GCP?

Reply:
That is not implemented. The "check_image_exists" implemented in the ecr class isn't actually used for anything yet though we should probably have the runners call this before they use an image.

Reply:
Okay, let's file a ticket for adding that verification step. IMO we shouldn't ship this method if its unused, we can always pull it in when we implement the ticket. This keeps code clutter out.

Reply:
done :)